src/celestia/celx_misc.cpp

```diff
@@ -261,132 +241,109 @@ static int image_new(lua_State* l, Image* i)
+    return celx.push(s.data());
 }
 
 void CreateImageMetaTable(lua_State* l)
 {
-    Celx_CreateClassMetatable(l, Celx_Image);
-
-    Celx_RegisterMethod(l, "__tostring", image_tostring);
-    Celx_RegisterMethod(l, "getheight", image_getheight);
-    Celx_RegisterMethod(l, "getwidth", image_getwidth);
-
-    lua_pop(l, 1); // remove metatable from stack
-}
-
-// ==================== Texture ============================================
-
-int texture_new(lua_State* l, Texture* t)
-{
-    Texture** ud = static_cast<Texture**>(lua_newuserdata(l, sizeof(Texture*)));
-    *ud = t;
-
-    Celx_SetClass(l, Celx_Texture);
+    CelxLua celx(l);
 
-    return 1;
-}
+    celx.createClassMetatable(Celx_Image);
 
-static Texture* to_texture(lua_State* l, int index)
-{
-    Texture** texture = static_cast<Texture**>(lua_touserdata(l, index));
+    celx.registerMethod("__tostring", image_tostring);
+    celx.registerMethod("getheight", image_getheight);
+    celx.registerMethod("getwidth", image_getwidth);
 
-    return texture != nullptr ? *texture : nullptr;
+    celx.pop(1); // remove metatable from stack
 }
 
-static Texture* this_texture(lua_State* l)
-{
-    Texture* texture = to_texture(l,1);
-    if (texture == nullptr)
-        Celx_DoError(l, "Bad texture object!");
+// ==================== Texture ============================================
 
-    return texture;
-}
+#define this_tex getThis<Texture*>()
 
 static int texture_bind(lua_State* l)
 {
-    Celx_CheckArgs(l, 1, 1, "No arguments expected for texture:bind()");
+    CelxLua celx(l);
 
-    Texture* texture = this_texture(l);
+    celx.checkArgs(1, 1, "No arguments expected for texture:bind()");
+
+    auto texture = *celx.this_tex;
     texture->bind();
     return 0;
 }
 
 static int texture_getheight(lua_State* l)
 {
-    Celx_CheckArgs(l, 1, 1, "No arguments expected for texture:getheight()");
+    CelxLua celx(l);
 
-    Texture* texture = this_texture(l);
-    lua_pushnumber(l, texture->getHeight());
-    return 1;
+    celx.checkArgs(1, 1, "No arguments expected for texture:getheight()");
+
+    auto texture = *celx.this_tex;
+    return celx.push(texture->getHeight());
 }
 
 static int texture_getwidth(lua_State* l)
 {
-    Celx_CheckArgs(l, 1, 1, "No arguments expected for texture:getwidth()");
+    CelxLua celx(l);
 
-    Texture* texture = this_texture(l);
-    lua_pushnumber(l, texture->getWidth());
-    return 1;
+    celx.checkArgs(1, 1, "No arguments expected for texture:getwidth()");
+
+    auto texture = *celx.this_tex;
+    return celx.push(texture->getWidth());
 }
 
 static int texture_tostring(lua_State* l)
 {
-    // TODO: print out the actual information about the texture
-    lua_pushstring(l, "[Texture]");
+    CelxLua celx(l);
 
-    return 1;
+    auto tex = *celx.this_tex;
+    std::string s = fmt::sprintf("[Texture:%dx%d]", tex->getWidth(), tex->getHeight());
+    return celx.push(s.data());
```

Comment:
s.c_str()

Reply:
What's wrong with `data()`? According to documentation, it's identical to `c_str()`.

Reply:
when you see c_str() you understand that receiver wants char*, what does it want if you see data() - void*?
and Celestia uses c_str() exclusively. let's stick to one style